wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -798,27 +800,41 @@ private String sessionUserNameOr(String nameIfNoUser) {
     return nameIfNoUser;
   }
 
-  private class AbnormalRuntimeStopListener implements EventSubscriber<RuntimeStatusEvent> {
+  private class AbnormalRuntimeStoppingListener
+      implements EventSubscriber<RuntimeAbnormalStoppingEvent> {
 
     @Override
-    public void onEvent(RuntimeStatusEvent event) {
-      if (event.isFailed()) {
-        String workspaceId = event.getIdentity().getWorkspaceId();
-        // Cancels workspace servers probes if any
-        probeScheduler.cancel(workspaceId);
-        final WorkspaceStatus status;
-        try (Unlocker ignored = lockService.writeLock(workspaceId)) {
-          runtimes.remove(workspaceId);
-          status = statuses.remove(workspaceId);
-        }
-        if (status != null) {
-          publishWorkspaceStatusEvent(
-              workspaceId,
-              STOPPED,
-              status,
-              "Error occurs on workspace runtime stop. Error: " + event.getError());
-          setAbnormalStopAttributes(workspaceId, event.getError());
-        }
+    public void onEvent(RuntimeAbnormalStoppingEvent event) {
+      String workspaceId = event.getIdentity().getWorkspaceId();
+      try (Unlocker ignored = lockService.writeLock(workspaceId)) {
+        statuses.replace(workspaceId, STOPPING);
+      }
+
+      publishWorkspaceStatusEvent(
+          workspaceId,
+          STOPPING,
+          event.getPreviousStatus(),
+          "Workspace is going to be STOPPED. Reason: " + event.getReason());
+    }
+  }
+
+  private class AbnormalRuntimeStoppedListener
+      implements EventSubscriber<RuntimeAbnormalStoppedEvent> {
+
+    @Override
+    public void onEvent(RuntimeAbnormalStoppedEvent event) {
+      String workspaceId = event.getIdentity().getWorkspaceId();
+      // Cancels workspace servers probes if any
+      probeScheduler.cancel(workspaceId);
+      final WorkspaceStatus status;
+      try (Unlocker ignored = lockService.writeLock(workspaceId)) {
+        runtimes.remove(workspaceId);
+        status = statuses.remove(workspaceId);
+      }
+      if (status != null) {
```

Comment:
It is not clear for me too =)
I added this check just to avoid NPE. I guess it should not be null here but still.
I'll take a look this place again and add the corresponding comment

Reply:
I think it MUST NOT be null when an event is received. I've added logging if status was not cached because potentially it may be a bug. See https://github.com/eclipse/che/pull/11570/commits/00a041f2b8a20b0ab7fcd786296e68bb6d5a4c4c